File: back/src/main/java/br/com/eletrodescarte/models/Descarte.java
package br.com.eletrodescarte.models;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
@Entity
@Table(name = "descartes")
public class Descarte {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_descarte")
    private Long idDescarte;

    @CreationTimestamp
    @Column(name = "descartado_em", updatable = false)
    private LocalDateTime descartadoEm;

    private String observacoes;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_usuario", nullable = false)
    private Usuario usuario;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_ponto", nullable = false)
    private PontoColeta pontoColeta;

    @OneToMany(mappedBy = "descarte", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<DescarteItem> itens;
}

File: mobile/lib/models/descarte.dart
import 'usuario.dart';
import 'ponto_coleta.dart';
import 'material_model.dart';

class DescarteItem {
  final int? idItem;
  final double quantidadeKg;
  final MaterialModel material;

  DescarteItem({
    this.idItem,
    required this.quantidadeKg,
    required this.material,
  });

  factory DescarteItem.fromJson(Map<String, dynamic> json) {
    return DescarteItem(
      idItem: json['idItem'] ?? json['id_item'],
      quantidadeKg: (json['quantidadeKg'] ?? json['quantidade_kg'] ?? 0.0) as double,
      material: MaterialModel.fromJson(json['material']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (idItem != null) 'idItem': idItem,
      'quantidadeKg': quantidadeKg,
      'material': {
        'idMaterial': material.idMaterial,
      },
    };
  }
}

class Descarte {
  final int? idDescarte;
  final String? descartadoEm;
  final String? observacoes;
  final Usuario usuario;
  final PontoColeta pontoColeta;
  final List<DescarteItem> itens;

  Descarte({
    this.idDescarte,
    this.descartadoEm,
    this.observacoes,
    required this.usuario,
    required this.pontoColeta,
    required this.itens,
  });

  factory Descarte.fromJson(Map<String, dynamic> json) {
    var listItens = json['itens'] as List? ?? [];
    List<DescarteItem> itensList = listItens
        .map((i) => DescarteItem.fromJson(i))
        .toList();

    return Descarte(
      idDescarte: json['idDescarte'] ?? json['id_descarte'],
      descartadoEm: json['descartadoEm'] ?? json['descartado_em'],
      observacoes: json['observacoes'],
      usuario: Usuario.fromJson(json['usuario']),
      pontoColeta: PontoColeta.fromJson(json['pontoColeta']),
      itens: itensList,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (idDescarte != null) 'idDescarte': idDescarte,
      if (descartadoEm != null) 'descartadoEm': descartadoEm,
      'observacoes': observacoes ?? '',
      'usuario': {
        'idUsuario': usuario.idUsuario,
      },
      'pontoColeta': {
        'idPonto': pontoColeta.idPonto,
      },
      'itens': itens.map((i) => i.toJson()).toList(),
    };
  }
}
